subsys/net/lib/http/http_server.c

```diff
@@ -0,0 +1,1182 @@
+/*
+ * Copyright (c) 2023, Emna Rekik
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include "headers/http_service.h"
+#include "headers/server_functions.h"
+
+#include <errno.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <strings.h>
+
+#include <zephyr/data/json.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/net/http/service.h>
+#include <zephyr/net/tls_credentials.h>
+LOG_MODULE_REGISTER(net_http_server, LOG_LEVEL_DBG);
+
+#define HTTP_SERVER_MAX_RESPONSE_SIZE CONFIG_NET_HTTP_SERVER_MAX_RESPONSE_SIZE
+#define HTTP_SERVER_MAX_URL_LENGTH    CONFIG_NET_HTTP_SERVER_MAX_URL_LENGTH
+#define HTTP_SERVER_MAX_FRAME_SIZE    CONFIG_NET_HTTP_SERVER_MAX_FRAME_SIZE
+
+#if !defined(__ZEPHYR__) || defined(CONFIG_POSIX_API)
+
+#define INCLUDE_HTML_CONTENT 1
+
+#include <arpa/inet.h>
+#include <netdb.h>
+#include <netinet/in.h>
+#include <poll.h>
+#include <sys/eventfd.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#else
+
+#include <zephyr/kernel.h>
+#include <zephyr/net/net_ip.h>
+#include <zephyr/net/socket.h>
+#include <zephyr/posix/sys/eventfd.h>
+
+#endif
+
+static char url_buffer[CONFIG_NET_HTTP_SERVER_MAX_URL_LENGTH];
+static char http_response[CONFIG_NET_HTTP_SERVER_MAX_RESPONSE_SIZE];
+static const char *preface = "PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n";
+
+static struct http_parser_settings parserSettings;
+static struct http_parser parser;
+
+static unsigned char settings_frame[9] = {
+	0x00, 0x00, 0x00,        /* Length */
+	0x04,                    /* Type: 0x04 - setting frames for config or acknowledgment */
+	0x00,                    /* Flags: 0x00 - unused flags */
+	0x00, 0x00, 0x00, 0x00}; /* Reserved, Stream Identifier: 0x00 - overall connection */
+
+static unsigned char settings_ack[9] = {
+	0x00, 0x00, 0x00,        /* Length */
+	0x04,                    /* Type: 0x04 - setting frames for config or acknowledgment */
+	0x01,                    /* Flags: 0x01 - ACK */
+	0x00, 0x00, 0x00, 0x00}; /* Reserved, Stream Identifier */
+
+static const char content_404[] = {
+#ifdef INCLUDE_HTML_CONTENT
+#include "not_found_page.html.gz.inc"
+#endif
+};
+
+bool has_upgrade_header = true;
```

Comment:
this seems to be some state that should probably be part of an http request detail, rather than a global variable.